# Assignment: Data Wrangling

In this assignment, we will practice data wrangling techniques on real world data.

## NYC film permits

New York City is a very popular film/tv shooting location. In order to assert the exclusive usage of city properties (e.g. sidewalk, park, etc.) for this purpose, a permit is required. New York Mayor's Office of Media and Entertainment release all granted [film permits data](https://data.cityofnewyork.us/City-Government/Film-Permits/tg4x-b46p) to the public. In this part, we will do some data wrangling on this data.

Let us first retrieve the data!

In [1]:
import pandas as pd

url = "https://data.cityofnewyork.us/api/views/tg4x-b46p/rows.csv"
    
df = pd.read_csv(url)

Your task for this part is to clean up this data. Specifically, the cleaned up dataframe should have the following properties:
1. All columns must have the correct data type.  In particular, `StartDateTime`, `EndDateTime`, `EnteredOn` should have `datetime64` as data type.
2. The `ZipCode(s)` column may contain multiple zip codes in one cell. In the cleaned up dataframe, we will replace the `ZipCode(s)` column with `ZipCode` column, where each cell in the `ZipCode` column only contains a single zip code. Rows with multiple zip codes in the input data frame should be repeated for each zip code in the output dataframe.
3. The output dataframe should not contain any missing data.

In [2]:
df.head().T

0  \
EventID                                                       696255   
EventType                                            Shooting Permit   
StartDateTime                                 02/17/2023 09:00:00 AM   
EndDateTime                                   02/18/2023 12:00:00 PM   
EnteredOn                                     02/14/2023 10:47:33 PM   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        KINGSLAND AVENUE between DEAD END and GREENPOI...   
Borough                                                     Brooklyn   
CommunityBoard(s)                                                  1   
PolicePrecinct(s)                                                 94   
Category                                                  Television   
SubCategoryName                                       Cable-episodic   
Country                                     United States of America   
ZipCode(s)                                                     11222   

                                                                   1  \
EventID                                                       714139   
EventType                                            Shooting Permit   
StartDateTime                                 05/12/2023 01:00:00 PM   
EndDateTime                                   05/13/2023 05:00:00 AM   
EnteredOn                                     05/04/2023 02:27:51 PM   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        WEST   26 STREET between 12 AVENUE and 11 AVEN...   
Borough                                                    Manhattan   
CommunityBoard(s)                                                  4   
PolicePrecinct(s)                                                 10   
Category                                                  Television   
SubCategoryName                                      Episodic series   
Country                                     United States of America   
ZipCode(s)                                                     10001   

                                                                   2  \
EventID                                                       705334   
EventType                                            Shooting Permit   
StartDateTime                                 04/10/2023 09:00:00 AM   
EndDateTime                                   04/10/2023 10:00:00 PM   
EnteredOn                                     03/30/2023 05:17:29 PM   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        SOUTH STREET between BROAD STREET and OLD SLIP...   
Borough                                                    Manhattan   
CommunityBoard(s)                                               1, 3   
PolicePrecinct(s)                                               1, 5   
Category                                                        Film   
SubCategoryName                                              Feature   
Country                                     United States of America   
ZipCode(s)                                       10002, 10004, 10005   

                                                                   3  \
EventID                                                       746696   
EventType                                            Shooting Permit   
StartDateTime                                 11/07/2023 06:00:00 AM   
EndDateTime                                   11/07/2023 10:00:00 PM   
EnteredOn                                     10/31/2023 12:08:00 PM   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        NORTH HENRY STREET between GREENPOINT AVENUE a...   
Borough                                                     Brooklyn   
CommunityBoard(s)                                                  1   
PolicePrecinct(s)                                                 94   
Category                        

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7870 entries, 0 to 7869
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   EventID            7870 non-null   int64 
 1   EventType          7870 non-null   object
 2   StartDateTime      7870 non-null   object
 3   EndDateTime        7870 non-null   object
 4   EnteredOn          7870 non-null   object
 5   EventAgency        7870 non-null   object
 6   ParkingHeld        7870 non-null   object
 7   Borough            7870 non-null   object
 8   CommunityBoard(s)  7868 non-null   object
 9   PolicePrecinct(s)  7868 non-null   object
 10  Category           7870 non-null   object
 11  SubCategoryName    7870 non-null   object
 12  Country            7870 non-null   object
 13  ZipCode(s)         7868 non-null   object
dtypes: int64(1), object(13)
memory usage: 860.9+ KB


In [4]:
df["EventType"].value_counts()

Shooting Permit                  6160
Theater Load in and Load Outs    1512
Rigging Permit                    151
DCAS Prep/Shoot/Wrap Permit        47
Name: EventType, dtype: int64

In [5]:
df["StartDateTime"] = pd.to_datetime(df["StartDateTime"])
df["EndDateTime"] = pd.to_datetime(df["EndDateTime"])
df["EnteredOn"] = pd.to_datetime(df["EnteredOn"])

In [6]:
df["EventAgency"].value_counts()

Mayor's Office of Media & Entertainment    7870
Name: EventAgency, dtype: int64

In [7]:
len(df["ParkingHeld"].value_counts())

4214

In [8]:
df["Borough"].value_counts()

Manhattan        3904
Brooklyn         2517
Queens           1102
Bronx             306
Staten Island      41
Name: Borough, dtype: int64

In [9]:
df["CommunityBoard(s)"].value_counts()

1              1851
2              1042
5               909
4               539
7               436
               ... 
11, 2, 6, 8       1
3, 4, 8           1
12, 13            1
1, 3, 6, 7        1
14, 2, 6          1
Name: CommunityBoard(s), Length: 231, dtype: int64

In [10]:
len(df["PolicePrecinct(s)"].value_counts())

538

In [11]:
len(df["Category"].value_counts())

9

In [12]:
len(df["SubCategoryName"].value_counts())

26

In [13]:
df["ZipCode(s)"] = df["ZipCode(s)"].astype(str)

In [14]:
zip_df = df[df['ZipCode(s)'].str.contains(',')]
# zip_df.T

In [15]:
zip_df["ZipCode(s)"] = zip_df["ZipCode(s)"].astype(str)

/tmp/ipykernel_25019/1317475753.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  zip_df["ZipCode(s)"] = zip_df["ZipCode(s)"].astype(str)


In [16]:
expanded_rows = []
for index, row in zip_df.iterrows():
    zip_code_list = row["ZipCode(s)"].split(",")

    for zip_code in zip_code_list:
        new_row = row.copy()
        new_row["ZipCode(s)"] = zip_code.strip()
        
        expanded_rows.append(new_row)

expanded_df = pd.DataFrame(expanded_rows)


In [17]:
expanded_df.T

2     \
EventID                                                       705334   
EventType                                            Shooting Permit   
StartDateTime                                    2023-04-10 09:00:00   
EndDateTime                                      2023-04-10 22:00:00   
EnteredOn                                        2023-03-30 17:17:29   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        SOUTH STREET between BROAD STREET and OLD SLIP...   
Borough                                                    Manhattan   
CommunityBoard(s)                                               1, 3   
PolicePrecinct(s)                                               1, 5   
Category                                                        Film   
SubCategoryName                                              Feature   
Country                                     United States of America   
ZipCode(s)                                                     10002   

                                                                2     \
EventID                                                       705334   
EventType                                            Shooting Permit   
StartDateTime                                    2023-04-10 09:00:00   
EndDateTime                                      2023-04-10 22:00:00   
EnteredOn                                        2023-03-30 17:17:29   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        SOUTH STREET between BROAD STREET and OLD SLIP...   
Borough                                                    Manhattan   
CommunityBoard(s)                                               1, 3   
PolicePrecinct(s)                                               1, 5   
Category                                                        Film   
SubCategoryName                                              Feature   
Country                                     United States of America   
ZipCode(s)                                                     10004   

                                                                2     \
EventID                                                       705334   
EventType                                            Shooting Permit   
StartDateTime                                    2023-04-10 09:00:00   
EndDateTime                                      2023-04-10 22:00:00   
EnteredOn                                        2023-03-30 17:17:29   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        SOUTH STREET between BROAD STREET and OLD SLIP...   
Borough                                                    Manhattan   
CommunityBoard(s)                                               1, 3   
PolicePrecinct(s)                                               1, 5   
Category                                                        Film   
SubCategoryName                                              Feature   
Country                                     United States of America   
ZipCode(s)                                                     10005   

                                                                11    \
EventID                                                       744752   
EventType                                            Shooting Permit   
StartDateTime                                    2023-10-17 09:00:00   
EndDateTime                                      2023-10-17 23:45:00   
EnteredOn                                        2023-10-13 14:34:03   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        PIKE STREET between CHERRY STREET and MONROE S...   
Borough                                                    Manhattan   
CommunityBoard(s)                                               1, 3   
PolicePrecinct(s)                                           5, 7, 94   
Category                     

In [18]:
clean_df = df[~df['ZipCode(s)'].str.contains(',')]

In [19]:
clean_df.T

0     \
EventID                                                       696255   
EventType                                            Shooting Permit   
StartDateTime                                    2023-02-17 09:00:00   
EndDateTime                                      2023-02-18 12:00:00   
EnteredOn                                        2023-02-14 22:47:33   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        KINGSLAND AVENUE between DEAD END and GREENPOI...   
Borough                                                     Brooklyn   
CommunityBoard(s)                                                  1   
PolicePrecinct(s)                                                 94   
Category                                                  Television   
SubCategoryName                                       Cable-episodic   
Country                                     United States of America   
ZipCode(s)                                                     11222   

                                                                1     \
EventID                                                       714139   
EventType                                            Shooting Permit   
StartDateTime                                    2023-05-12 13:00:00   
EndDateTime                                      2023-05-13 05:00:00   
EnteredOn                                        2023-05-04 14:27:51   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        WEST   26 STREET between 12 AVENUE and 11 AVEN...   
Borough                                                    Manhattan   
CommunityBoard(s)                                                  4   
PolicePrecinct(s)                                                 10   
Category                                                  Television   
SubCategoryName                                      Episodic series   
Country                                     United States of America   
ZipCode(s)                                                     10001   

                                                                3     \
EventID                                                       746696   
EventType                                            Shooting Permit   
StartDateTime                                    2023-11-07 06:00:00   
EndDateTime                                      2023-11-07 22:00:00   
EnteredOn                                        2023-10-31 12:08:00   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        NORTH HENRY STREET between GREENPOINT AVENUE a...   
Borough                                                     Brooklyn   
CommunityBoard(s)                                                  1   
PolicePrecinct(s)                                                 94   
Category                                                  Commercial   
SubCategoryName                                           Commercial   
Country                                     United States of America   
ZipCode(s)                                                     11222   

                                                                4     \
EventID                                                       717328   
EventType                              Theater Load in and Load Outs   
StartDateTime                                    2023-05-31 00:01:00   
EndDateTime                                      2023-06-01 06:00:00   
EnteredOn                                        2023-05-17 11:55:51   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        WEST   55 STREET between 11 AVENUE and 12 AVEN...   
Borough                                                    Manhattan   
CommunityBoard(s)                                                  4   
PolicePrecinct(s)                                                 18   
Category                     

In [20]:
final_df = pd.concat ([clean_df, expanded_df], ignore_index=True)

In [21]:
final_df.T

0      \
EventID                                                       696255   
EventType                                            Shooting Permit   
StartDateTime                                    2023-02-17 09:00:00   
EndDateTime                                      2023-02-18 12:00:00   
EnteredOn                                        2023-02-14 22:47:33   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        KINGSLAND AVENUE between DEAD END and GREENPOI...   
Borough                                                     Brooklyn   
CommunityBoard(s)                                                  1   
PolicePrecinct(s)                                                 94   
Category                                                  Television   
SubCategoryName                                       Cable-episodic   
Country                                     United States of America   
ZipCode(s)                                                     11222   

                                                               1      \
EventID                                                       714139   
EventType                                            Shooting Permit   
StartDateTime                                    2023-05-12 13:00:00   
EndDateTime                                      2023-05-13 05:00:00   
EnteredOn                                        2023-05-04 14:27:51   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        WEST   26 STREET between 12 AVENUE and 11 AVEN...   
Borough                                                    Manhattan   
CommunityBoard(s)                                                  4   
PolicePrecinct(s)                                                 10   
Category                                                  Television   
SubCategoryName                                      Episodic series   
Country                                     United States of America   
ZipCode(s)                                                     10001   

                                                               2      \
EventID                                                       746696   
EventType                                            Shooting Permit   
StartDateTime                                    2023-11-07 06:00:00   
EndDateTime                                      2023-11-07 22:00:00   
EnteredOn                                        2023-10-31 12:08:00   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        NORTH HENRY STREET between GREENPOINT AVENUE a...   
Borough                                                     Brooklyn   
CommunityBoard(s)                                                  1   
PolicePrecinct(s)                                                 94   
Category                                                  Commercial   
SubCategoryName                                           Commercial   
Country                                     United States of America   
ZipCode(s)                                                     11222   

                                                               3      \
EventID                                                       717328   
EventType                              Theater Load in and Load Outs   
StartDateTime                                    2023-05-31 00:01:00   
EndDateTime                                      2023-06-01 06:00:00   
EnteredOn                                        2023-05-17 11:55:51   
EventAgency                  Mayor's Office of Media & Entertainment   
ParkingHeld        WEST   55 STREET between 11 AVENUE and 12 AVEN...   
Borough                                                    Manhattan   
CommunityBoard(s)                                                  4   
PolicePrecinct(s)                                                 18   
Category                    

In [22]:
final_df.isna().sum().sum()
final_df.dropna(inplace = True)

In [23]:
final_df.isna().sum().sum()

0

# Once the data frame is clean, please answer the following questions with code:
1. Which borough has the most _unique_ events for television, film and theater respectively?
2. Which zip code has the most _unique_ events for television, film and theater respectively?

In [24]:
final_df.columns

Index(['EventID', 'EventType', 'StartDateTime', 'EndDateTime', 'EnteredOn',
       'EventAgency', 'ParkingHeld', 'Borough', 'CommunityBoard(s)',
       'PolicePrecinct(s)', 'Category', 'SubCategoryName', 'Country',
       'ZipCode(s)'],
      dtype='object')

In [25]:
final_df["Category"].value_counts()


Television           5515
Theater              2149
Film                 1608
Commercial           1249
Still Photography     903
WEB                   535
Documentary            96
Student                53
Music Video            52
Name: Category, dtype: int64

In [26]:
# entertainment = final_df[final_df["Category"].isin(["Television", "Theater", "Film"])]

# # len(entertainment["EventID"].unique()
# # print (entertainment.shape)

# entertainment = entertainment.drop_duplicates(subset = ["EventID"])

# # entertainment.groupby("Borough")
# entertainment.groupby("Borough").count()

# # de = abc [abc["Category"] == "Brooklyn" ] 
# # entertainment.shape

# # group = entertainment.groupby("Borough")


In [27]:
entertainment = final_df[final_df["Category"].isin(["Television", "Theater", "Film"])]

entertainment = entertainment.drop_duplicates(subset = ["EventID"])

borough_event_counts = entertainment.groupby(["Borough", "Category"])["EventID"].nunique()

In [28]:
borough_event_counts

Borough        Category  
Bronx          Film            44
               Television     223
               Theater          3
Brooklyn       Film           360
               Television    1199
               Theater        343
Manhattan      Film           416
               Television    1310
               Theater       1190
Queens         Film           122
               Television     874
Staten Island  Film            28
               Television       5
Name: EventID, dtype: int64

In [29]:
boro_tv = borough_event_counts.xs("Television", level = "Category").idxmax()
tv_c= borough_event_counts.xs("Television", level = "Category").max()

boro_f = borough_event_counts.xs("Film", level = "Category").idxmax()
f_c= borough_event_counts.xs("Film", level = "Category").max()

boro_th = borough_event_counts.xs("Theater", level = "Category").idxmax()
th_c= borough_event_counts.xs("Theater", level = "Category").max()

print (f"Borough with highest number of TELEVISION events is {boro_tv} with {tv_c} events.")
print (f"Borough with highest number of FILM events is {boro_f} with {f_c} events.")
print (f"Borough with highest number of Theater events is {boro_th} with {th_c} events.")

Borough with highest number of TELEVISION events is Manhattan with 1310 events.
Borough with highest number of FILM events is Manhattan with 416 events.
Borough with highest number of Theater events is Manhattan with 1190 events.


In [30]:
entertainment2 = final_df[final_df["Category"].isin(["Television", "Film", "Theater"])]
entertainment2 = entertainment2.drop_duplicates(subset="EventID")

# entertainment2["ZipCode(s)"] = entertainment2["ZipCode(s)"].astype(int)
zip_event_counts = entertainment2.groupby(["ZipCode(s)", "Category"])["EventID"].nunique()



zip_event_counts

ZipCode(s)  Category  
0           Film           1
            Television     1
00083       Film           1
            Television     1
10001       Film          26
                          ..
11435       Television     4
11691       Television     1
11693       Television     1
11694       Film           4
            Television     2
Name: EventID, Length: 270, dtype: int64

In [31]:
zip_tv = zip_event_counts.xs("Television", level = "Category" ).idxmax()
tv_c2 = zip_event_counts.xs("Television", level = "Category" ).max()

zip_f = zip_event_counts.xs("Film", level = "Category" ).idxmax()
f_c2 = zip_event_counts.xs("Film", level = "Category" ).max()

zip_th = zip_event_counts.xs("Theater", level = "Category" ).idxmax()
th_c2 = zip_event_counts.xs("Theater", level = "Category" ).max()


In [32]:
print (f"Borough with highest number of TELEVISION events is {boro_tv} with {tv_c} events.")
print (f"Borough with highest number of FILM events is {boro_f} with {f_c} events.")
print (f"Borough with highest number of Theater events is {boro_th} with {th_c} events.")
print()

print (f"ZipCode with highest number of TELEVISION events is {zip_tv} with {tv_c2} events.")
print (f"ZipCode with highest number of FILM events is {zip_f} with {f_c2} events.")
print (f"ZipCode with highest number of Theater events is {zip_th} with {th_c2} events.")


Borough with highest number of TELEVISION events is Manhattan with 1310 events.
Borough with highest number of FILM events is Manhattan with 416 events.
Borough with highest number of Theater events is Manhattan with 1190 events.

ZipCode with highest number of TELEVISION events is 11222 with 778 events.
ZipCode with highest number of FILM events is 10002 with 71 events.
ZipCode with highest number of Theater events is 10019 with 295 events.


## Answer

### Student name: Rohan Gore (N19332535)

### Your answer:
1. The borough with the most unique TV events is ...
    Answer: Manhattan
   
2. The borough with the most unique film events is ...
    Answer: Manhattan

3. The borough with the most unique theater events is ...
    Answer: Manhattan


4. The zip code with the most unique TV events is ...
    Answer: 11222

5. The zip code with the most unique film events is ...
    Answer: 10002
 
6. The zip code with the most unique theater events is ...
    Answer: 10019


In [33]:
print("GG")

GG
